src/extensions/dataarray.js

```diff
@@ -0,0 +1,82 @@
+import { datastream } from '../constants';
+import express        from 'express';
+import response       from '../response';
+import * as ERR       from '../errors';
+
+/**
+ * Implementation of 13.2 "Create Observation entities with dataArray"
+ *
+ * Similar to the SWE DataArray in the OGC SOS, SensorThings API also
+ * provides the support of dataArray (in addition to formatting every
+ * observation entity as a JSON object) to aggregate multiple Observation
+ * entities and reduce the request (e.g., POST) and response (e.g., GET)
+ * size. SensorThings mainly use dataArray in two scenarios: (1) get
+ * Observation entities in dataArray, and (2) create Observation entities
+ * with dataArray.
+ *
+ **/
+
+export default version => {
+  return (req, res, next) => {
+
+    // Required properties in each object:
+    //   - Datastream
+    //   - phenomenonTime
+    //   - result
+    //   - featureOfInterest
+    let requiredProperties = ['phenomenonTime', 'result', 'FeatureOfInterest'];
+    var validateObject = function(obj) {
+      if (obj.Datastream === undefined) {
+        if (missingProperties.indexOf(datastream) === -1) {
+          missingProperties.push(datastream);
+        }
+      }
+
+      let properties = obj.components.toString();
+      requiredProperties.forEach(property => {
+        if (properties.indexOf(property) == -1) {
+          if (missingProperties.indexOf(property) === -1) {
+            missingProperties.push(property);
+          }
+        }
+      });
+    };
+
+    var dump = function(obj) {
+        for (var prop in obj) {
+          if (typeof obj[prop] === "object" && Object.prototype.toString.call(obj[prop]) !== '[object Array]') {
+            dump(obj[prop]);
+          } else {
+            console.log('obj.' + prop + ' = ' + obj[prop]);
+          }
+        }
+    };
+
+    console.log('request body as JSON:', req.body);
+
+    let body = req.body;
+
+    // Validate input.
+    // 1) The body of request must be an array.
+    if (Object.prototype.toString.call(body) !== '[object Array]') {
+      ERR.ApiError(res, 400, ERR.ERRNO_BAD_REQUEST,
+                   ERR.BAD_REQUEST, 'Input must be an array');
+    }
+
+    // 2) Valildate that each object in the input has the required properties.
+    let missingProperties = [];
```

Comment:
I also wasn't sure if it was worth the effort to check the entire request. And now I've decided it is not worth the effort so the code now bails when it detects the first missing property. This makes the code more straightforward and efficient.